mtp_api/apps/mtp_auth/serializers.py

```diff
@@ -181,11 +178,18 @@ def create(self, validated_data):
         new_user.set_password(password)
         new_user.save()
 
-        if make_user_admin:
+        role.assign_to_user(new_user)
+        # In custom flow, enforce that `UserAdmin` is added to any new `Security` user who has the `can_manage` box
+        # ticked in the form.
+        # Note that the django admin user creation flow does not use this endpoint so this logic won't apply there
+        if make_user_admin or new_user.groups.filter(name='FIU').exists():
```

Comment:
I'm reading this and I'm wondering if this `if` statement could be split into two `if` statements to make things more clear.

AFAICT (and I may well be wrong!), if `make_user_admin=True` then we want to add this user to the `UserAdmin` group - this is basically the logic which was there before.

On top of that, if the user is part of `FIU` group then we also want to add them to the `UserAdmin` group.

And finally, if someone is in the `Security` group and can manage users (is in the `UserAdmin` group) that means they should be added to the `FIU` group, that seems to be implicit by the fact this `if` is nested within this `if` - but again, maybe I'm misunderstanding this.

What I'm wondering and potentially suggest is something like:

```Python
# FIU users can manage users
if new_user.groups.filter(name='FIU').exists():
  make_user_admin = True

if make_user_admin:
  new_user.groups.add(Group.objects.get(name='UserAdmin'))

# Security users that can manage users should also be in the FIU group
if make_user_admin and new_user.groups.filter(name='Security').exists():
  new_user.groups.add(Group.objects.get(name='UserAdmin'))
```

I suppose it's the same logic but maybe it's a bit easier to follow 🤷

Reply:
i'm not sure that's clearer. bearing in mind that the driving input to decide whether somebody is an admin _or_ FIU is the `user_admin=True` from the POST request. to me, it's more logical to have that drive the decision to add both FIU and UserAdmin groups

Reply:
what i mean is, to me it's clearer if only the POST request sets `make_user_admin` and this variable doesn't then change. but it's subjective, right? it's only about readability rather than overall effect.

Reply:
(Thinking aloud here) Actually I also notice that this logic to add/remove a 'Security' user to/from the 'FIU' group is repeated in several places and I do wonder if it would beneficial to move into a function to sync those two groups.

Maybe something like this:

```Python
def sync_security_fiu_groups(user):
    """
    Add/remove a 'Security' user to/from the 'FIU' group.

    If a user is in the 'Security' group and can manage users they should
    also be in the 'FIU' group.

    Conversely, if a 'Security' user can't manage users they should be
    removed from the 'FIU' group.
    """

    can_manage_user = updated_user.groups.filter(name='UserAdmin').exists()
    is_security = updated_user.groups.filter(name='Security').exists()

    if is_security:
        if can_manage_user:
            user.groups.add(Group.objects.filter(name='FIU').first())
        else:
            user.groups.remove(Group.objects.filter(name='FIU').first())
```

You could then simplify some of the code below with something like:

```python
if make_user_admin:
  updated_user.groups.add(user_admin_group)
else:
  updated_user.groups.remove(user_admin_group)

sync_security_fiu_groups(updated_user)
```
(or something like that, all of this is pretty much pseudo-code and I may miss a lot of things here)

but maybe it's overkill and it would hide this logic and make these create/update methods more difficult to read 🤷

Reply:
yeah that works for me, but i have no strong opinions either way :)

Reply:
@ushkarev 

> i'm not sure this code snippet is clearer. bearing in mind that the driving input to decide whether somebody is an admin or FIU is the user_admin=True from the POST request. to me, it's more logical to have that drive the decision to add both FIU and UserAdmin groups

Yes, I see what you mean about the input variable being repurposed 😬 - it was a bit of an idea and wrote that way to avoid the `or` in the first `if ` and/or to avoid the repetition in of the `new_user.groups.add()`, you may well write that like:

```python
# FIU users can also manage users
if make_user_admin or new_user.groups.filter(name='FIU').exists():
  new_user.groups.add(Group.objects.get(name='UserAdmin'))

# Security users that can manage users should also be in the FIU group
if make_user_admin and new_user.groups.filter(name='Security').exists():
  new_user.groups.add(Group.objects.get(name='UserAdmin'))
```

I suppose the thing I struggle a little to read was the nested if to add a `Security` user to the `FIU` group but maybe it's just me have a preference with flat `if` statements without and/or (whenever it's possible of course) ☺️

Reply:
we definitely cannot have 

```python
if make_user_admin and new_user.groups.filter(name='Security').exists():
  new_user.groups.add(Group.objects.get(name='UserAdmin'))
```

because then no users outside of the security tool can be made an admin

Reply:
Yep 👍, sorry that was a copy&paste mistake which I noticed when I re-read my comment, it was meant to be add to the 'FIU' group clearly ☺️

Reply:
So i'm happy to break it out into functions if it makes sense to do so. 

Couple of points worth making though:

Regarding your first comment, I think having nested if statements in this case is clearer than having shallow if statements predicated on two conditions. Also I'm not sure if  redefining a value we're receiving as input to the form adds to the clarity of the code; as Igor says I think it's clearer to use the notion of the `user_admin` provided by the form here. Also the `FIU` group wouldn't be ever present on the user at this point; as FIU doesn't have an existing role, it fits in the existing functionality much better if we think of them as admins of the `Security` role, so that the only change we need to make to the generic code flow as it stands is to make sure that any user created with `role=security` and `make_admin=True` has the `FIU` group added to them, rather than the other way round (see my comment https://dsdmoj.atlassian.net/browse/MTP-1818?focusedCommentId=143381)

> ```python
> def sync_security_fiu_groups(user):
>     """
>     Add/remove a 'Security' user to/from the 'FIU' group.
> 
>     If a user is in the 'Security' group and can manage users they should
>     also be in the 'FIU' group.
> 
>     Conversely, if a 'Security' user can't manage users they should be
>     removed from the 'FIU' group.
>     """
> 
>     can_manage_user = updated_user.groups.filter(name='UserAdmin').exists()
>     is_security = updated_user.groups.filter(name='Security').exists()
> 
>     if is_security:
>         if can_manage_user:
>             user.groups.add(Group.objects.filter(name='FIU').first())
>         else:
>             user.groups.remove(Group.objects.filter(name='FIU').first())
> ```

I think this last line will raise an exception when updating a normal `Security` user, as the `FIU` group won't be part of that users group set

Reply:
i agree with you @Staberinde 

on a side note, `user.groups.remove(Group.objects.filter(name='FIU').first())` can be turned into `user.groups.filter(name='FIU').remove()` (or is it .delete()?) so that it doesn't raise exceptions